dali/kernels/imgproc/pointwise/linear_transformation_gpu.h

```diff
@@ -0,0 +1,195 @@
+class LinearTransformationGpu {
+ private:
+  static constexpr auto ndims_ = spatial_ndims + 1;
+  using Mat = ::dali::mat<channels_out, channels_in, float>;
+  using Vec = ::dali::vec<channels_out, float>;
+  using BlockDesc = kernels::BlockDesc<spatial_ndims>;
+  using SampleDescriptor = lin_trans::SampleDescriptor<OutputType, InputType,
+          channels_out, channels_in, spatial_ndims>;
+
+  std::vector<SampleDescriptor> sample_descriptors_;
+
+ public:
+  BlockSetup<spatial_ndims,
+          spatial_ndims /* Assumed, that the channel dim is the last dim */> block_setup_;
+
+
+  KernelRequirements
+  Setup(KernelContext &context, const InListGPU<InputType, ndims_> &in, span<const Mat> tmatrices,
+        span<const Vec> tvectors, span<const Roi<spatial_ndims>> rois = {}) {
+    DALI_ENFORCE(rois.empty() || rois.size() == static_cast<size_t>(in.num_samples()),
+                 "Provide ROIs either for all or none input tensors");
+    for (int i = 0; i < in.size(); i++) {
+      DALI_ENFORCE(in[i].shape.shape.back() == channels_in,
+                   make_string("Unexpected number of channels at index", i,
+                               "in InListGPU. Number of channels in every InListGPU has to match"
+                               " the number of channels, that the kernel is instantiated with"));
+    }
+    for (size_t i = 0; i < rois.size(); i++) {
+      DALI_ENFORCE(all_coords(rois[i].hi >= rois[i].lo),
+                   make_string("Found invalid ROI at index", i,
+                               "ROI doesn't follow {lo, hi} convention.", rois[i]));
+    }
+
+    auto adjusted_rois = AdjustRoi(rois, in.shape);
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    TensorListShape<ndims_> output_shape(ShapeFromRoi(make_cspan(adjusted_rois), channels_out));
+    block_setup_.SetupBlocks(output_shape, true);
+    se.add<SampleDescriptor>(AllocType::GPU, in.num_samples());
+    se.add<BlockDesc>(AllocType::GPU, block_setup_.Blocks().size());
+    req.output_shapes = {output_shape};
+    req.scratch_sizes = se.sizes;
+    return req;
+  }
+
+
+  void Run(KernelContext &context, const OutListGPU<OutputType, spatial_ndims + 1> &out,
+           const InListGPU<InputType, spatial_ndims + 1> &in, span<const Mat> tmatrices,
+           span<const Vec> tvectors, span<const Roi<spatial_ndims>> rois = {}) {
+    auto sample_descs = lin_trans::CreateSampleDescriptors(out, in, tmatrices, tvectors, rois);
+
+    typename decltype(sample_descs)::value_type *samples_gpu;
+    BlockDesc *blocks_gpu;
+
+    std::tie(samples_gpu, blocks_gpu) = context.scratchpad->ToContiguousGPU(
+            context.gpu.stream, sample_descs, block_setup_.Blocks());
+
+    dim3 grid_dim = block_setup_.GridDim();
+    dim3 block_dim = block_setup_.BlockDim();
+    auto stream = context.gpu.stream;
+    // @autoformat:off
+    lin_trans::LinearTransformationKernel
+            <OutputType, InputType, channels_out, channels_in, spatial_ndims>
```

Comment:
Could you give it a try and remove this line? Maybe after changing mat/vec to `int ndim` the compiler will be able to deduce all template arguments.

Reply:
Done